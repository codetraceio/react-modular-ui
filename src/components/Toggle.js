import React from 'react';
import AbstractComponent from './AbstractComponent';

export default class Toggle extends AbstractComponent {
  getModifierObject() {
    return {
      size: this.props.size,
      color: this.props.color,
      checked: this.props.checked,
      disabled: this.props.disabled
    };
  }

  isChecked() {
    return this.props.checked || false;
  }

  onClick(e) {
    this.props.onChange(!this.props.checked, e);
  }

  render() {
    return (
      <div
        className={this.getBlockName('toggle', this.getModifierObject())}
        role="checkbox"
        tabIndex="0"
        aria-checked={this.isChecked().toString()}
        onClick={(e) => this.onClick(e)}
      >
        <div className={this.getElementName('toggle', 'container')}>
          <div className={this.getElementName('toggle', 'line')} />
          <div className={this.getElementName('toggle', 'handle')} />
        </div>
        {this.props.children ? (
          <div className={this.getElementName('toggle', 'label')}>
            {this.props.children}
          </div>
        ) :  null}
      </div>
    )
  }
}
