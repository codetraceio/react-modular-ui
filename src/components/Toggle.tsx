import React, { PropsWithChildren, useCallback, useContext } from "react";

import { className } from "../utils/className";
import { ThemeContext } from "./ThemeContext";

export interface ToggleProps {
  size?: string | number;
  color?: string;
  disabled?: boolean;
  checked?: boolean;
  theme?: string;

  onChange?: (value: boolean, event: React.MouseEvent<HTMLDivElement>) => void;
}

export default function Toggle({
  size,
  color,
  disabled,
  checked,
  theme,
  children,
  onChange,
  ...props
}: PropsWithChildren<ToggleProps>) {
  const themeContext = useContext(ThemeContext);

  const handleClick = useCallback((e: React.MouseEvent<HTMLDivElement>) => {
    if (typeof onChange === "function") {
      onChange(!checked, e);
    }
  }, [checked, onChange]);

  return (
    <div
      className={className("toggle")}
      role="checkbox"
      tabIndex={0}
      aria-disabled={disabled}
      aria-checked={checked}
      data-size={size}
      data-color={color}
      data-theme={theme ?? themeContext}

      onClick={handleClick}
      {...props}
    >
      <div className={className("toggle", "container")}>
        <div className={className("toggle", "line")} />
        <div className={className("toggle", "handle")} />
      </div>
      {children ? (
        <div className={className("toggle", "label")}>
          {children}
        </div>
      ) :  null}
    </div>
  );
}
