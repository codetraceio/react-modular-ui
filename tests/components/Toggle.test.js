import React from 'react';
import renderer from 'react-test-renderer';
import Toggle from '../../src/components/Toggle';

describe('Toggle', () => {
  it('should render Toggle', () => {
    const tree = renderer.create(
      <Toggle
        size="32"
        view="flat"
        color="primary"
        disabled
      >Something</Toggle>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
