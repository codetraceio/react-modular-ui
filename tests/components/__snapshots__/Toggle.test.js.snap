// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle should render Toggle 1`] = `
<div
  aria-checked="false"
  className="c-toggle -size-32 -color-primary -disabled"
  onClick={[Function]}
  role="checkbox"
  tabIndex="0"
>
  <div
    className="c-toggle--container"
  >
    <div
      className="c-toggle--line"
    />
    <div
      className="c-toggle--handle"
    />
  </div>
  <div
    className="c-toggle--label"
  >
    Something
  </div>
</div>
`;
